crates/bevy_render/src/view/window/screenshot.rs

```diff
@@ -221,114 +496,179 @@ impl SpecializedRenderPipeline for ScreenshotToScreenPipeline {
     }
 }
 
-pub struct ScreenshotPreparedState {
-    pub texture: Texture,
-    pub buffer: Buffer,
-    pub bind_group: BindGroup,
-    pub pipeline_id: CachedRenderPipelineId,
-}
-
 pub(crate) fn submit_screenshot_commands(world: &World, encoder: &mut CommandEncoder) {
-    let windows = world.resource::<ExtractedWindows>();
+    let targets = world.resource::<RenderScreenshotTargets>();
+    let prepared = world.resource::<RenderScreenshotsPrepared>();
     let pipelines = world.resource::<PipelineCache>();
-
-    for window in windows.values() {
-        if let Some(memory) = &window.screenshot_memory {
-            let width = window.physical_width;
-            let height = window.physical_height;
-            let texture_format = window.swap_chain_texture_format.unwrap();
-
-            encoder.copy_texture_to_buffer(
-                memory.texture.as_image_copy(),
-                wgpu::ImageCopyBuffer {
-                    buffer: &memory.buffer,
-                    layout: layout_data(width, height, texture_format),
-                },
-                Extent3d {
-                    width,
-                    height,
-                    ..Default::default()
-                },
-            );
-            if let Some(pipeline) = pipelines.get_render_pipeline(memory.pipeline_id) {
-                let true_swapchain_texture_view = window
+    let gpu_images = world.resource::<RenderAssets<GpuImage>>();
+    let windows = world.resource::<ExtractedWindows>();
+    let manual_texture_views = world.resource::<ManualTextureViews>();
+
+    for (entity, render_target) in targets.iter() {
+        match render_target {
+            NormalizedRenderTarget::Window(window) => {
+                let window = window.entity();
+                let window = windows.get(&window).unwrap();
+                let width = window.physical_width;
+                let height = window.physical_height;
+                let texture_format = window.swap_chain_texture_format.unwrap();
```

Comment:
Is this ever None? Or do we already know this is window always that has this?

Reply:
The scenario in which it's `None` is if you supply a `RenderTarget` with a window ref that's an invalid window entity. I made it so we don't extract those but I'll guard here just to be safe, unwraps are always a bit sketchy.

Reply:
Cleaned up a ton of unwraps which should be safe but u never know.